### PR TITLE
c8d/mount: Use ref-counted mounter by default

### LINKED ISSUE #46842: c8d: unconditionally use ref-counting, or get info from snapshotter
### Description

- relates to https://github.com/moby/moby/pull/45698
- relates to https://github.com/moby/moby/pull/45781
- relates to https://github.com/moby/moby/pull/45781#discussion_r1235917043
- slightly related to https://github.com/moby/moby/issues/46757

Commit https://github.com/moby/moby/commit/21c0a54a6b2dc059fae62f46a23d338a35fe6d1d (https://github.com/moby/moby/pull/45698) introduced a hard-coded list of snapshotters (`refCountedFileSystems`) that we know need reference-counting. With containerd, the list of available snapshotters is not known (additional snapshotters may be installed on the system), so maintaining a hard-coded list is problematic to maintain.

https://github.com/moby/moby/blob/54fcd40aa4de94cd75aedc5f6ebf38c6d8f92082/daemon/snapshotter/mount.go#L15-L16

We should consider to either;

- unconditionally use reference-counting (as suggested in https://github.com/moby/moby/pull/45781#discussion_r1236612727)
- get this information from the snapshotter (also related: https://github.com/moby/moby/issues/46757)
- :point_up: this may depend on containerd 2.0 though, so perhaps we need to go for the alternative in the meantime

### PR DESCRIPTION
**Pull request by vvoland**
- closes: https://github.com/moby/moby/issues/46842

All commonly used filesystems should use ref-counted mounter, so make it the default instead of having to whitelist them.

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/snapshotter/mount.go [hunk 1]
```diff
@@ -13,8 +13,8 @@ import (
 	"github.com/moby/sys/mountinfo"
 )
 
-// List of known filesystems that can't be re-mounted or have shared layers
-var refCountedFileSystems = []string{"fuse-overlayfs", "overlayfs", "stargz", "zfs"}
+// List of known filesystems that can be re-mounted
+var notRefCountedFileSystems = []string{}
```
Comment (rumpl) [inline daemon/snapshotter/mount.go:17]: Why not completely remove since it's not used?

Comment (vvoland) [reply daemon/snapshotter/mount.go:17]: We could!

I assume that if you're asking, you don't recall any filesystem that specifically required the non-refcounted mounter? 😄

Comment (rumpl) [reply daemon/snapshotter/mount.go:17]: Exactly

**Comment by TBBle** [comment]
I did see it fail in 

```
=== Failed
=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references/local_volume_with_mount_options (2.34s)
    daemon_test.go:581: assertion failed: error is not nil: Error response from daemon: remove test-live-restore-volume-references-local: volume has active mounts
        --- FAIL: TestLiveRestore/volume_references/local_volume_with_mount_options (2.34s)

=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references (16.91s)
    --- FAIL: TestLiveRestore/volume_references (16.91s)

=== FAIL: amd64.integration.daemon TestLiveRestore (0.00s)
```

but only in [test (graphdriver) / integration (ubuntu-20.04, rootless)](https://github.com/moby/moby/actions/runs/7570683298/job/20616570235?pr=47105#logs).

I have no idea if that's a flake, or happens to be using the filesystem that needs to not be ref-counted, and triggers it because of how it's using mount options. Different mount options seem like something that should produce different mounts, and I don't see any code in `func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (target string, retErr error)` to ensure we don't accidentally inc-ref when passed read-write `mounts` when previously called with read-only `mounts`. Maybe that's not possible by usage, but the API appears to allow it.

File: daemon/snapshotter/mount.go (hunk 1 above)
Comment (vvoland) [reply daemon/snapshotter/mount.go]: Done.

**Review by rumpl — APPROVED** [review]

**Review by dmcgowan — APPROVED** [review]

**Pull request merged**